<html> <head>
<title>R/qtl software for mapping quantitative trait loci</title>

<meta name="description" content="R/qtl software for mapping
quantitative trait loci (QTL) in experimental crosses.">

<meta name="keywords" content="software, statistics, bioinformatics,
genetics, statistical genetics, QTL, quantitative trait loci, QTLs,
complex traits, quantitative trait locus, genetic map, QTL mapping">

<link rel=stylesheet type="text/css" href="kbroman.css">
<link rel="SHORTCUT ICON" href="r.ico">
</head>

<body>

<h2 align="center">R/qtl: A QTL mapping environment</h2>

<h3 align="center">Software for mapping quantitative trait loci in
experimental crosses</h3>

<p align="center"> Current version: 1.44-9 (2019-01-22)  <br>

<p align="center">[ <a
href="download">Download</a> | <a
href="faq">FAQ</a> | <a
href="https://github.com/kbroman/qtl/blob/master/NEWS.md">News</a> | <a
href="BUGS.txt">Bugs</a> | <a href="graphs">Sample
graphics</a> |
<a href="sampledata">Sample data</a> |
<a href="tutorials">Tutorials</a> |
<a href="book">Book</a> |
<a href="manual/qtl-manual.pdf">Manual</a> | <a
href="citation.html">Citation</a> ]

<p class=noindent> <hr>

<p class=center>Try the
<a
href="http://kbroman.org/qtlcharts">R/qtlcharts package</a>:
interactive graphics for QTL data.

<p class=noindent> <hr>

<p class=center><b>Check out our book:</b> <a href="book">A Guide to QTL Mapping with R/qtl</a>, by Karl W. Broman and
&#346;aunak Sen.</p>

<p class=noindent> <hr>

<p class=noindent> <table noborder cellpadding="0" cellspacing="1" width="100%">
<tr>
  <td align="center" width="48%"><h3>About
  R/qtl</h3>
  <td width="4%">
  <td align="center" width="48%"><h3>About
  R</h3>
  <tr> <td height=15>
<tr>
<td valign="top" align="left" width="48%">

  <p class=noindent>R/qtl is an extensible, interactive environment for mapping
  quantitative trait loci (QTL) in experimental crosses.   It is
  implemented as an add-on package for the freely available and widely
  used statistical language/software R (see the <a
  href="http://www.R-project.org" >R project
  homepage</a>).  The development of this software as an add-on to R
  allows us to take advantage of the basic mathematical and
  statistical functions, and powerful graphics capabilities, that are
  provided with R.  Further, the user will benefit by the seamless
  integration of the QTL mapping software into a general statistical
  analysis program.  Our goal is to make complex QTL mapping methods
  widely accessible and allow users to focus on modeling rather than
  computing.

  <p class=noindent>A key component of computational methods for QTL mapping is the
  hidden Markov model (HMM) technology for dealing with missing
  genotype data.  We have implemented the main HMM algorithms, with
  allowance for the presence of genotyping errors, for backcrosses,
  intercrosses, and phase-known four-way crosses.

  <p class=noindent>The current version of R/qtl includes facilities for estimating
  genetic maps, identifying genotyping errors, and performing
  single-QTL genome scans and two-QTL, two-dimensional genome scans,
  by interval mapping (with the EM algorithm), Haley-Knott regression,
  and multiple imputation.  All of this may be done in the presence of
  covariates (such as sex, age or treatment).  One may also fit higher-order
  QTL models by multiple imputation and Haley-Knott regression.

  <p class=noindent>R/qtl is distributed as source code for unix or compiled code for
  Windows or Mac.  R/qtl is released under the <a
  href="http://www.gnu.org/copyleft/gpl.html" >GNU
  General Public License</a>.  To download the software, you must agree
  to the terms in that download.

<td width="4%">
<td valign="top" align="left" width="48%">
<p class=noindent> R is an open-source implementation of the S language.  As described on
the <a href="http://www.R-project.org/" >R project
homepage</a>:

<p class=noindent> "<font color="#23238e">R is a system for statistical computation and
graphics.  It consists of a language plus a run-time environment with
graphics, a debugger, access to certain system functions, and the
ability to run programs stored in script files.</font>

<p class=noindent>"<font color="#23238e">The core of R is an interpreted computer language
which allows branching and looping as well as modular programming
using functions. Most of the user-visible functions in R are written
in R. It is possible for the user to interface to procedures written
in the C, C++, or FORTRAN languages for efficiency. The R distribution
contains functionality for a large number of statistical
procedures. Among these are: linear and generalized linear models,
nonlinear regression models, time series analysis, classical
parametric and nonparametric tests, clustering and smoothing. There is
also a large set of functions which provide a flexible graphical
environment for creating various kinds of data
presentations. Additional modules are available for a variety of
specific purposes.</font>"


<p class=noindent> R is freely available for Windows, unix and MacOS,
  and may be downloaded from the <a
  href="http://cran.r-project.org">Comprehensive R Archive Network</a>
  (CRAN).

<p class=noindent>  Learning R may require a formidable investment
  of time, but it will definitely be worth the effort.  Numerous free
  documents on getting started with R are available on
  <a href="http://cran.r-project.org">CRAN</a>.  In
  addition, several books are available on R, S and S-PLUS;
  for example, see WN Venables, BD Ripley (2002) <a
href="https://www.amazon.com/gp/product/0387954570?ie=UTF8&tag=7210-20">Modern Applied Statistics with S</a> (4th
ed, Springer) or P Dalgaard (2008) <a
  href="https://www.amazon.com/gp/product/0387790535?ie=UTF8&tag=7210-20">Introductory
  statistics with R</a> (2nd ed, Springer).

  <p class=noindent> See my <a href="http://www.biostat.wisc.edu/~kbroman/Rintro">Introduction to R</a> page for further
  links.

</table>

<p class=noindent> <hr>

<p class=noindent> <b>Contact</b> for
problems/questions/suggestions: <a href="http://kbroman.org">Karl W Broman</a>

<p class=noindent> <b>Authors</b>: Karl W Broman and Hao Wu, with ideas from Gary Churchill and &#346;aunak Sen and contributions from Danny Arends, Timoth&eacute;e Flutre, Ritsert Jansen, Pjotr Prins, Lars R&ouml;nneg&aring;rd, Rohan Shah, Laura Shannon, Quoc Tran, Aaron Wolen, and Brian Yandell

<p class=noindent><hr>
<p class=noindent> <b>Google Groups</b>:  We've created two Google
Groups for email announcements regarding software updates (<a
href="http://groups.google.com/group/Rqtl-announce">R/qtl announcements</a>)
and for discussion about the use of the software (<a
href="http://groups.google.com/group/Rqtl-disc">R/qtl discussion</a>).  Note that you
should join just one of these two groups; all announcements will also
be sent to the discussion group.



<p class=noindent> <hr>

<p class=noindent> <h3>Other QTL mapping software</h3>
<p class=noindent> <table noborder cellpadding="0" cellspacing="1" width="100%">

<!--
    <a href="http://www.mapmanager.org/mmQTX.html">Map Manager QTX</a>
-->

<tr><td width="4%">
<td valign="top" width="48%">
    <a href="http://www.ssg.uab.edu/qtlbim">R/qtlbim</a> (runs in R)
<td valign="top" width="48%">
<a href="http://ibi.zju.edu.cn/software/qtlnetwork/index.htm">QTLNetwork</a>

<tr><td>
<td valign="top" width="48%">
  <a href="http://churchill.jax.org/software/archive/pseudomarker.shtml">Pseudomarker</a> (runs in Matlab)
<td valign="top" width="48%">
    <a href="http://www.genenetwork.org/home.html">WebQTL</a>

<tr><td>
<td valign="top" width="48%">
  <a href="http://www.broad.mit.edu/ftp/distribution/software/mapmaker3/">MapMaker/QTL</a>
<td valign="top" width="48%">
  <a  href="http://www.multiqtl.com">MultiQTL</a>

<tr><td>
<td valign="top" width="48%">
  <a href="http://statgen.ncsu.edu/qtlcart/cartographer.html">QTL Cartographer</a>
<td valign="top" width="48%">
  <a href="http://www.uni-hohenheim.de/~ipspwww/soft.html">PLABQTL</a>

<tr><td>
<td valign="top" width="48%">
  <a href="http://www.epibiostat.ucsf.edu/biostat/sen/qtldesign-software/">R/qtlDesign</a>
<td valign="top" width="48%">
  <a href="http://www.biosciences-labs.bham.ac.uk/Kearsey/applet.html">The QTL Cafe</a>

<tr><td>
<td valign="top" width="48%">
  <a href="http://www.well.ox.ac.uk/~rmott/happy.html">HAPPY</a>
<td valign="top" width="48%">
  <a href="http://www.gridqtl.org.uk">GridQTL</a>

<tr><td>
<td valign="top" width="48%">
  <a href="http://www.kyazma.nl/index.php/mc.MapQTL">MapQTL</a>
<td valign="top" width="48%">
  <a href="http://user.it.uu.se/~kl/qtl_software.html">Kajsa Ljungberg's software</a>

<tr><td>
<td valign="top" width="48%">
  <a href="http://famprevmed.ucsd.edu/faculty/cberry/bqtl/">BQTL</a> (runs in R)
<td valign="top" width="48%">
  <a href="https://forge-dga.jouy.inra.fr/projects/qtlmap">QTLMap</a></td>

<tr><td>
<td valign="top" width="48%">
  <a href="http://www.RNI.Helsinki.FI/~mjs/">Multimapper</a>
<td valign="top" width="48%">
  <!-- blank -->
</table>

<p class=noindent> <hr>

<p class=noindent> <h3>Also of interest</h3>

<p class=noindent> <table noborder cellpadding="0" cellspacing="1" width="100%">
<tr><td width="4%">
<td valign="top" width="48%">
Gary
Churchill's <a href="http://www.qtlarchive.org">QTL Archive</a></td>
<td valign="top" width="48%"></td></tr>
</table>


<p class=noindent> <hr>


<p align="center">[ <a
href="download">Download</a> | <a
href="faq">FAQ</a> | <a
href="https://github.com/kbroman/qtl/blob/master/NEWS.md">News</a> | <a
href="BUGS.txt">Bugs</a> | <a href="graphs">Sample
graphics</a> |
<a href="sampledata">Sample data</a> |
<a href="tutorials">Tutorials</a> |
<a href="book">Book</a> |
<a href="manual/qtl-manual.pdf">Manual</a> | <a
href="citation.html">Citation</a> ]

<hr>
<p class=noindent>
<table cellspacing="0" cellpadding="0" width=100%>
<tr><td align="left">
<a href="http://kbroman.org">Karl Broman's home page</a>

</body></html>
