<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8" />

<title>CTC 2013: Poster 28 (Laura Saba)</title>

<link rel="stylesheet" href="../ctc2013.css" type="text/css" />
</head>

<body>

<p>Poster 28<br>
<em>Presenter</em>: Laura Saba<br>
<a href="../program_thu.html">Thursday, 4:00 &ndash; 6:00pm</a><h3>Identifying transcriptional signatures of brain region-specific volume from whole brain RNA-Seq data</h3>

<p><strong>Laura M. Saba<sup>1</sup>, Robert W. Williams<sup>2</sup>, Ashutosh Pandey<sup>2</sup>, Paula L. Hoffman<sup>1</sup>, Boris Tabakoff<sup>1</sup></strong></br>

<em><sup>1</sup>Department of Pharmacology, University of Colorado School of Medicine, Aurora CO USA and <sup>2</sup>Department of Anatomy and Neurobiology, University of Tennessee Health Science Center, Memphis TN USA</em></p>

<p>For many heritable complex neurological phenotypes, such as substance use disorders (SUD), no single brain region or cell type is solely responsible for its etiology and it is clear that the brain operates as a network of functionally linked cells and regions.  Often in such situations of ambiguity, whole brain tissue is used for initial transcriptome analyses. However, transcriptional associations in brain can be confounded by genetic variation in proportions of brain regions and cell types, and these differences can create confusion about the source of transcriptional variances (e.g., transcriptional differences within cells or differences in the proportion of cells).  We evaluated weighted gene co-expression network analysis (WGCNA) and independent component analysis (ICA) based on their ability to extract signals related to brain region-specific volume from whole brain transcriptome data from the BXD recombinant inbred panel (http://www.genenetwork.org).  These signals allow volume to be accounted for in associations between brain transcription levels and phenotypes.  Over 8,000 highly expressed transcripts were identified and quantified using RNA-Seq data from 31 RI strains. WGCNA and ICA identified eigengenes and source signals, respectively, that were associated with differences in proportional brain region volumes.  ICA had the additional benefit of relating the different sources back to individual transcripts that were expressed in more than one brain region.  ICA is useful for extracting expression signals from multiple sources and including these signals in transcriptional analyses of complex traits will increase interpretability of genetic associations. Supported by NIAAA (AA013162, AA013162-08S1, AA016662, AA013499), the NFPCDD, and the Banbury Fund.</p>

