<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8" />

<title>CTC 2013: Poster 41 (Beth Wilmot)</title>

<link rel="stylesheet" href="../ctc2013.css" type="text/css" />
</head>

<body>

<p>Poster 41<br>
<em>Presenter</em>: Beth Wilmot<br>
<a href="../program_wed.html">Wednesday, 3:00 &ndash; 5:00pm</a><h3>Reading  the Whole Transcriptome</h3>

<p><strong>Sunita Kawane<sup>1</sup>, Christina L. Zheng<sup>2,3</sup>, Daniel W. Bottomly<sup>1,3</sup>, Robert Searles<sup>4</sup>, Robert Hitzemann<sup>5,7</sup>, Shannon McWeeney<sup>1,2,3,6</sup>, Beth Wilmot<sup>1,2,3</sup></strong></br>

<em><sup>1</sup>Oregon Clinical and Translational Research Institute, <sup>2</sup> Department of Medical Informatics and Clinical Epidemiology, <sup>3</sup>the Knight Cancer Center, <sup>4</sup>Integrated Genomics Laboratory, <sup>5</sup>Department of Behavioral Neuroscience, <sup>6</sup>Department of Public Health and Preventative Medicine, Oregon Health &amp; Science University, Portland, Oregon, <sup>7</sup>Veterans Affairs Research Service, Portland.</em></p>

<p>In order to utilize the full power of RNAseq, these two aspects of transcriptomics that need to be addressed: 1) assigning reads to the approximately 10% of overlapping genes and 2) developing a full annotation of reads aligning to non-exonic regions.  We have developed a framework that allows investigation of both the complex regions of genic overlap and annotation of reads that align to unannotated regions of the genome.  We used stranded libraries of polyA RNA or riboZero treated RNA in the development and testing of this framework. Scenarios when each of these protocols would be appropriate are discussed to guide discovery.  In the resulting annotation files, the genome is portioned into categories according to strand, exon and intron overlaps and intergenic regions.  This is summarized at the gene, exon, transcript and intron/intergenic regions.  Complex regions of overlapping genes were further characterized.  To help define regions where reads align to nongenic areas of the genome, Non-Code annotation and other genomic features such as histone marks and DHS can also be integrated. This is highly extensible for any RNA-seq pipeline and freely available to the research community.  This work was supported by grants AA010760, AA011034, MH051372, and AA013484.</p>

