<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8" />

<title>CTC 2013: Poster 21 (Stephen Flink)</title>

<link rel="stylesheet" href="../ctc2013.css" type="text/css" />
</head>

<body>

<p>Poster 21<br>
<em>Presenter</em>: Stephen Flink<br>
<a href="../program_wed.html">Wednesday, 3:00 &ndash; 5:00pm</a><h3>Genomic variants in the parental strains of the rat HxB/BxH recombinant inbred panel</h3>

<p><strong>Stephen Flink<sup>1</sup>, Laura M. Saba<sup>1</sup>, Morton Printz<sup>2</sup>, Laura Breen, Paula L. Hoffman<sup>1</sup>, Boris Tabakoff<sup>1</sup></strong></br>

<em><sup>1</sup>Department of Pharmacology, University of Colorado School of Medicine, Aurora, CO 80045 <sup>2</sup>Department of Pharmacology, University of California San Diego, La Jolla, CA, USA <sup>3</sup>National Institute for Cellular Biotechnology, Dublin City University, Dublin <sup>9</sup>, Ireland</em></p>

<p>Panels of recombinant inbred strains are an invaluable tool for genetical/genomic analyses of complex traits.  Continuing refinements in genome and transcriptome sequencing provide increasingly detailed sets of markers  for genome-wide analyses including differences in RNA expression responsible for phenotypic diversity.<p>The parental strains of the HxB/BxH rat RI panel are the spontaneously hypertensive rat strain, SHR/OlaIpcv, and the BN-Lx/Cub, a Brown Norway congenic strain with polydactyl-luxate syndrome.  These two strains are believed to be among the most genetically diverse among laboratory rat strains in general use, and show physiological and behavioral differences beyond those traits for which they were originally bred.<p>We sequenced the DNA of the BN-Lx/Cub and SHR/OlaIpcv strains and mapped the resulting sequences to the latest Brown Norway (BN) reference genome (RGSC 5.0/rn5).  We identified over 3.2 million single-nucleotide polymorphisms and over 600,000 small  (&lt;15 nt) genomic insertions or deletions between the parental strains using a standard samtools/bcftools pipeline.  We also identified larger genomic variants including insertions and deletions of length â‰¥15nt, inversions, and differences in copy number between the parental strains and the reference genome using more novel analysis methods. Many of the variants between the strains may have a significant impact on protein production and function, with as many as 50,000 small variants occurring within or near protein-coding genes. Data from this study are available at our website: http://phenogen.ucdenver.edu/PhenoGen/ Supported by NIAAA (AA013162, AA013162-08S1) and the Banbury Fund.</p>

