<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8" />

<title>CTC 2013: Poster 38 (Christina Zheng)</title>

<link rel="stylesheet" href="../ctc2013.css" type="text/css" />
</head>

<body>

<p>Poster 38<br>
<em>Presenter</em>: Christina Zheng<br>
<a href="../program_thu.html">Thursday, 4:00 &ndash; 6:00pm</a><h3>Splicing Landscape of 8 Founder Strains</h3>

<p><strong>Christina L. Zheng<sup>1,2</sup>, Beth Wilmot<sup>1,2,3</sup>, Sunita Kawane<sup>3</sup>, Robert P. Searles<sup>4</sup>, Shannon McWeeney<sup>1,2,3,5</sup>, Robert Hitzemann<sup>6,7</sup></strong></br>

<em><sup>1</sup>Department of Medical Informatics and Clinical Epidemiology, <sup>2</sup>Knight Cancer Institute, <sup>3</sup>Oregon Clinical and Translational Research Institute, <sup>4</sup>Integrated Genomics Laboratory, <sup>5</sup>Department of Public Health and Preventative Medicine, <sup>6</sup>Department of Behavioral Neuroscience, Oregon Health &amp; Science University, Portland, Oregon, <sup>7</sup>Veterans Affairs Research Service, Portland, OR.</em></p>

<p>Inbred laboratory mouse strains are an invaluable research tool.   In an effort to understand the functional differences among different strains, we have analyzed the splicing landscape of eight inbred mouse strains.  The strains include those of classical laboratory strains (129, A/J, C57, NOD, and NZO) and 3 wild derived inbred strains (CAST, PWK and WSB).  Using RNA-seq data mapped to the C57BL/6J mouse reference genome, we found that much of the splicing landscape across the strains are shared with ~70% of all spliced junctions being shared betwween at least 2 strains and 24% of the junctions being shared among all the strains.  However ~10-16% of junctions were found to be strain specific, with PWK and CAST having the highest percentages.  A striking majority (~96%) of the strain specific junctions were found to be novel unannotated junctions.  Furthermore among the high confidence strain specific junctions (&gt;  read coverage of 10) we found that only ~10% of them resided within annotated genes with ~94% of the genes being unique to an individual strain.  These findings suggest that key splicing differences may help to further define the functional differences between the strains. The resulting strain specific junctions (genomic coordinates and read coverage information) is freely available to the research community.  This work was supported by grants AA010760, AA011034, MH051372, and AA013484.</p>

