<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8" />

<title>CTC 2013: Poster 42 (Greg Carter)</title>

<link rel="stylesheet" href="../ctc2013.css" type="text/css" />
</head>

<body>

<p>Poster 42<br>
<em>Presenter</em>: Greg Carter<br>
<a href="../program_thu.html">Thursday, 4:00 &ndash; 6:00pm</a><h3>Deciphering Genetic Complexity with the Combined Analysis of Pleiotropy and Epistasis</h3>

<p><strong>Gregory W. Carter</strong></br>

<em>The Jackson Laboratory</em></p>

<p>Recent advances in mouse genetic resources, high-resolution genotyping, and multidimensional phenotyping are designed to enable precise genetic modeling of complex biological systems. The success of this approach is contingent upon the continued development of computational methods to dissect genetic complexity. Here we present an analytical strategy to infer models of how multiple genetic variants interact to influence multiple phenotypes. The method combines signals of epistasis between partially pleiotropic genes across multiple phenotypes to derive specific models of how each genetic variant enhances or suppresses the effects of other variants, and, in turn, affects each phenotype. The resulting network model interprets statistical epistasis in terms of more specific, directional hypotheses of variant-to-variant action. The method is designed to be flexible and scalable for application to populations with extensive genetic diversity. We present examples in yeast, fly, and mouse model systems, demonstrating the ability of the approach to both map large-scale genetic architecture and generate specific pair-wise genetic hypotheses.</p>

