<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8" />

<title>CTC 2013: Talk 13 (Beverly Mock)</title>

<link rel="stylesheet" href="../ctc2013.css" type="text/css" />
</head>

<body>

<p>Talk 13<br>
<em>Presenter</em>: Beverly Mock<br>
<a href="../program_thu.html">Thursday, 10:40am</a></p>

<h3>Biologically-Integrated Network Analyses Reveal Synergistic Action of A Drug Combination Targeting Cancer Susceptibility Pathways</h3>

<p><strong>John K. Simmons<sup>1*</sup>, Aleksandra M. Michalowski<sup>1*</sup>, Benjamin Gamache<sup>1</sup>, Jyoti B. Patel<sup>1</sup>, Adriana Zingone<sup>2</sup>, Ke Zhang<sup>1</sup>, Michael Kuehl<sup>3</sup>, Jing Huang<sup>1</sup>, Ola Landgren<sup>2</sup>, and Beverly A. Mock<sup>1</sup></strong></br>

<em><sup>1</sup>Lab of Cancer Biology and Genetics; <sup>2</sup>Metabolism Branch; <sup>3</sup>Genetics Branch, CCR, NCI, NIH, Bethesda, MD (<sup>*</sup>equal contribution)</em></p>

<p>Synergistic drug combinations have proven effective in cancer therapy. Defining drug synergy at the molecular level has proven challenging, yet, identification of synergy relevant response targets could be used to understand mechanism of combination action and define treatment responsive patient subsets. The combination of entinostat (Class I HDACi) and sirolimus (mTORi) was used to target pathways underlying genetic predispositions identified in a mouse model of myeloma. Weighted gene co-expression network analysis (WGCNA) of gene expression profile data from a human multiple myeloma (MM) cell line treated with single agents and their combination was used to identify a distinct module of 126 genes cooperatively affected by the drug combination; 37 of these genes were found to be differentially expressed in plasma cells from normal donors vs. MM patients, and were predictive of survival by multivariate analyses. Protein and mRNA target validations were performed. Pharmacodynamic responses of the signature to the drug combination were confirmed (Nanostring) in several cell lines from multiple tumor types and in ex vivo-treated primary patient samples before and after treatment.  Ingenuity transcription factor (TF) analysis predicted 6 TFs upstream of the 37 responsive genes; CHIP-seq dataset mining confirmed that two TFs bound the promoters of 32 of the 37 genes.  Further experiments have been performed to elucidate functional links between the two TFs, the 37 gene signature, and the drug response.</p>

