<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8" />

<title>CTC 2013: Talk 5 (Stuart Macdonald)</title>

<link rel="stylesheet" href="../ctc2013.css" type="text/css" />
</head>

<body>

<p>Talk 5<br>
<em>Presenter</em>: Stuart Macdonald<br>
<a href="../program_wed.html">Wednesday, 11:40am</a></p>

<h3>The Drosophila Synthetic Population Resource</h3>

<p><strong>Stuart J Macdonald<sup>*</sup>, Elizabeth G King<sup>**</sup>, Casey L McNeil<sup>*</sup>, Jennifer L Hackett<sup>*</sup>, Sophia S Loschky<sup>*</sup>, Brittny R Smith<sup>*</sup>, Michael A Najarro<sup>*</sup>, Tara N Marriage<sup>*</sup>, Anthony D Long<sup>**</sup></strong></br>

<em><sup>*</sup>Department of Molecular Biosciences, University of Kansas. <sup>**</sup>Department of Ecology and Evolutionary Biology, University of California, Irvine.</em></p>

<p>Genetic dissection of complex, polygenic trait variation is a key goal of medical and evolutionary genetics. Attempts to identify variants underlying complex traits have been plagued by low mapping resolution in traditional linkage studies, and an inability to identify variants that cumulatively explain the bulk of standing genetic variation in genomewide association studies (GWAS). We have developed a novel resource for the Drosophila community consisting of two sets of recombinant inbred lines (RILs), each derived from an advanced generation intercross between a different set of eight highly inbred, completely resequenced founders. The Drosophila Synthetic Population Resource (DSPR) has been designed to combine the high mapping resolution offered by multiple generations of recombination, with the high statistical power afforded by a linkage-based design. We have assayed the 1,700 RILs for a range of traits, including drug-resistance phenotypes, mapping a large number of quantitative trait loci (QTL) to small intervals (&lt;0.5Mb). Many of the QTL we identify are rare - the minor allele is unique to a single founder line, and several do not show a simple biallelic pattern, suggesting multiple causative factors may frequently underlie QTL. To further characterize QTL contributing to a single trait - starvation resistance - we dissected the trait using three additional mapping designs. Over 3,000 heterozygous genotypes were generated by crossing pairs of RILs, and by independently backcrossing RILs to two isogenic tester strains. This work revealed a number of novel, potentially background-dependent resistance loci, and a surprisingly complex architecture for this important life history trait.</p>

