<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8" />

<title>CTC 2013: Talk 22 (Daniel Gatti)</title>

<link rel="stylesheet" href="../ctc2013.css" type="text/css" />
</head>

<body>

<p>Talk 22<br>
<em>Presenter</em>: Daniel Gatti<br>
<a href="../program_fri.html">Friday, 10:40am</a></p>

<h3>Quantitative Trait Locus Mapping in Diversity Outbred Mice</h3>

<p><strong>Daniel M. Gatti, Karen L. Svenson, Andrey Shabalin, Daniel Pomp, Neal Goodwin, Karl W. Broman, Gary A. Churchill</strong></br>

<em>The Jackson Laboratory, Bar Harbor, ME, USA; Medical College of Virginia of Virginia Commonwealth University, Richmond, VA, USA; Dept. of Genetics, University of North Carolina, Chapel Hill, NC, USA; Dept. of Biostatistics and Medical Informatics, University of Wisconsin, Madison, WI, USA</em></p>

<p>The search for genes underlying complex phenotypes has been greatly aided by genetic mapping in the mouse.  Traditionally, mapping has been carried out in two parent intercrosses with limited mapping resolution.  Relatively few of these studies have led directly to the discovery of a gene that regulates the phenotype of interest.  In order to improve mapping resolution, advanced intercrosses and multi-founder crosses have been developed.  Diversity Outbred (DO) mice were developed to overcome these limitations by combining high genetic diversity and fine recombination block structure in order to increase the chances of mapping a phenotype to a small region.  The DO mice are derived from the same set of eight founder strains as the Collaborative Cross and are maintained as an outbred population.  The task of reconstructing DO genomes and mapping requires specialized methods and software.  Here, we describe software uses a hidden Markov model to provide a probabilistic reconstruction of individual DO genomes from intensity based analysis of genotyping microarray data.  Genotype probabilities are used to map phenotypes in a mixed model that adjusts for the kinship among DO mice.  The model outputs additive effects for each founder allele that can be used to reduce the number of candidate genes under a mapping peak.  We provide a complete analytical pipeline, implemented as a freely available R package, to go from phenotypes and genotypes to candidate gene list.</p>

