<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8" />

<title>CTC 2013: Talk 21 (Steven Munger)</title>

<link rel="stylesheet" href="../ctc2013.css" type="text/css" />
</head>

<body>

<p>Talk 21<br>
<em>Presenter</em>: Steven Munger<br>
<a href="../program_fri.html">Friday, 10:20am</a></p>

<h3>RNA-seq alignment to individualized genomes.</h3>

<p><strong>Steven C. Munger<sup>1*</sup>, Narayanan Raghupathy<sup>1</sup>, Kwangbom Choi<sup>1</sup>, Allen K. Simons<sup>1</sup>, Daniel M. Gatti<sup>1</sup>, Douglas A. Hinerfeld<sup>1</sup>, Karen L. Svenson<sup>1</sup>, Mark P. Keller<sup>2</sup>, Alan D. Attie<sup>2</sup>, Matthew A. Hibbs<sup>1,3</sup>, Joel H. Graber<sup>1</sup>, Gary A. Churchill<sup>1</sup>, and Elissa J. Chesler<sup>1</sup></strong></br>

<em><sup>*</sup>Presenting author. <sup>1</sup>The Jackson Laboratory, Bar Harbor, ME 04609 <sup>2</sup>Department of Biochemistry, University of Wisconsin, Madison, WI 53706 <sup>3</sup>Department of Computer Science, Trinity University, San Antonio, TX 78212</em></p>

<p>Genetic variation that deviates from a reference genome sequence can create biases in the alignment of short sequencing reads and can distort transcript abundance estimates in RNA sequencing (RNA-seq). Error tolerant alignment does not correct this problem. We hypothesized that alignment of sequence reads to the imputed genomes of individuals (individualized genomes) from which tissue samples were obtained would reduce alignment errors and improve transcript quantification. We developed a method, implemented as the software package Seqnature, that constructs individualized genomes of experimental model organisms including inbred mouse strains and genetically unique outbred animals. We show that alignment of mouse liver RNA-seq reads to individualized genomes increases read mapping accuracy and improves transcript abundance estimates. In an application to expression QTL mapping in a large population of genetically heterogeneous Diversity Outbred (DO) mice, this approach corrected erroneous linkages and unmasked thousands of hidden local associations. Individualized genomes can be extended to human short-read sequencing and other sequencing applications including ChIP-seq.</p>

