<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8" />

<title>CTC 2013: Talk 8 (Leslie Turner)</title>

<link rel="stylesheet" href="../ctc2013.css" type="text/css" />
</head>

<body>

<p>Talk 8<br>
<em>Presenter</em>: Leslie Turner<br>
<a href="../program_wed.html">Wednesday, 1:40pm</a></p>

<h3>Genomic networks of hybrid sterility</h3>

<p><strong>Leslie M. Turner, Michael A. White, Diethard Tautz, Bret A. Payseur</strong></br>

<em>University of Wisconsin - Madison, Max Planck Institute for Evolutionary Biology</em></p>

<p>Hybrid dysfunction, a common feature of reproductive barriers between species, is often caused by negative epistasis between loci (Dobzhansky-Muller incompatibilities). Identifying both (or multiple) interaction partners is challenging using traditional genetic approaches, hence little is known about the nature and complexity of hybrid incompatibilities. Male hybrids between recently diverged subspecies of house mice (Mus musculus) often show reduced fertility. Previous studies have identified quantitative trait loci on the X and autosomes that contribute to phenotypes associated with sterility, and several X-autosome interactions. We used a systems genetics approach to identify disruptions in gene networks associated with sterility. We collected genome-wide testis expression data from 305 male F2s from a cross between wild-derived inbred strains of M. musculus musculus and M. m. domesticus. We identified several thousand cis- and trans-acting quantitative trait loci (eQTL) contributing to expression variation. Using a conditional mapping approach, we identified eQTL dependent on interactions between loci, revealing a complex pattern of epistasis. Many trans eQTL cluster into ten ‘hotspots,’ seven of which co-localize with QTL for sterility phenotypes identified in this cross. Functional annotation of transcripts with eQTL provides insights into the biological processes that are disrupted in sterile hybrids and guides prioritization of candidate genes. The systems genetics approach we employed is applicable in a broad range of organisms and we advocate for widespread adoption of a network-centered approach in speciation genetics.</p>

