<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8" />

<title>CTC 2013: Talk 3 (Beverly Richards-Smith)</title>

<link rel="stylesheet" href="../ctc2013.css" type="text/css" />
</head>

<body>

<p>Talk 3<br>
<em>Presenter</em>: Beverly Richards-Smith<br>
<a href="../program_wed.html">Wednesday, 11:00am</a></p>

<h3>What's in a Name? Why Nomenclature Matters</h3>

<p><strong>Beverly Richards-Smith, Howard Dene, Monica McAndrews, Judith Blake and Janan Eppig</strong></br>

<em>Mouse Genome Informatics, The Jackson Laboratory, Bar Harbor ME 04609, USA</em></p>

<p>Successful scientific exchange, through publication or discourse, assumes common understanding of the object(s) of study. For biologists, this means clear and unambiguous identity for genes, genetic/genomic variants and mutations, and genetically defined strains. Naming based on laboratory shorthand, field-of-study jargon, or common use continues to impede search effectiveness in repositories and resources where nomenclature standards are not enforced.<p>Nomenclature committees, such as those for mouse (International Committee on Standardized Genetic Nomenclature for Mice), rat (Rat Genome and Nomenclature Committee), and human (HUGO Gene Nomenclature Committee), work closely to harmonize ortholog naming among these well-studied species. Databases for these and other organisms, including ZFIN (zebrafish), FlyBase (Drosophila), SGD (yeast), and others, provide nomenclature guidance for their respective species.<p>Whole genome sequencing and assembly refinement over the years have established the number of protein-coding genes in human and mouse at 25-30,000. Discovery of new functional RNAs, a large array of regulatory elements, and highly conserved non-coding regions continues to reveal the complexity of the mammalian genome. Nomenclature continues to adapt to encompass these genomic elements and novel technology-enabled gene and chromosome modifications. Genetic background, particularly defined strains, on which specific mutations, quantitative trait loci (QTL), or transgenes are analyzed, comprises the totality of genetic inputs contributing to specific phenotypes. <p>We present examples of the benefits of standardized nomenclatures and how curation and application of nomenclature standards facilitate data integration and searching across online resources. Official Nomenclature Guidelines for mouse genes, alleles, chromosomal aberrations and strains are accessible from MGIâ€™s Nomenclature Page (www.informatics.jax.org/nomen). <p>Supported by NIH grant HG000330</p>

