<html> <head>
<title>Sample graphics from R/qtl</title>
<link rel=stylesheet type="text/css" href="../kbroman.css">
<link rel="SHORTCUT ICON" href="../r.ico">
</head>

<body>

<h2 align="center">Sample graphics from R/qtl</h2>

<p align="center">[ <a href="..">Home</a>
| <a href="../download">Download</a> | <a 
href="../faq">FAQ</a> | <a
href="../STATUS.txt">Status</a> | <a
href="../BUGS.txt">Bugs</a> |
<a href="../sampledata">Sample data</a> |
<a href="../tutorials">Tutorial</a> | 
<a href="http://www.rqtl.org/book">Book</a> |
<a href="../manual/qtl-manual.pdf">Manual</a> | <a
href="../citation.html">Citation</a> ] 

<p class=noindent><hr>

<p align="center">
<table align="center" noborder cellpadding="8" cellspacing="8"
  width="80%">
<tr> <td width="30%" align="right"> <a href="plot.cross.html"><img
     src="thumb_plot.cross.jpg" alt="Summary plot of experimental
  cross [69k]"></a>
     <td width="70%" align="left" valign="top"><b>Summary plot for an experimental
     cross</b>.  The top left panel displays the pattern of missing
     genotype data; the top right shows the genetic map.  The lower
     panels are histograms of two phenotypes. <font size=1></a>

<tr> <td align="right"> <a href="plot.missing.html"><img
     src="thumb_plot.missing.jpg" alt="Plot of missing genotypes [76k]"></a>
     <td valign="top"> <b>Pattern of missing genotype data in a cross</b>.  Black
     pixels indicate missing genotypes.  Gray pixels indicate
     partially missing genotypes (in an intercross, genotypes coded as
     "not B" or "not A"). 

<tr> <td align="right"> <a href="plot.map_1.html"><img
     src="thumb_plot.map_1.jpg" alt="Plot of genetic map [40k]"></a> 
     <td valign="top"> <b>Locations of markers on a genetic
  map</b>. 

<tr> <td align="right"> <a href="plot.map_2.html"><img
     src="thumb_plot.map_2.jpg" alt="Plot of comparison of two genetic
  maps [54k]"></a> 
     <td valign="top"> <b>Side-by-side comparison of two genetic maps</b>.  This
     can be useful for comparing the genetic map obtained from a
     database to that estimated with the observed data. 

<tr> <td align="right"> <a href="plot.rf.html"><img
     src="thumb_plot.rf.jpg" alt="Plot of recombination fractions [110k]"></a> 
     <td valign="top"> <b>Estimated recombination fractions</b> for all pairs of markers
     (on selected chromosomes), along with LOD scores for the test of
     r = 1/2.  (The recombination fractions are in the upper left
     triangle; LOD scores are in the lower right triangle.)  Red
     indicates a large LOD score or a small recombination fraction,
     while blue is the opposite.  Note that a marker on chromosome 8
     appears to be switched with a marker on chromosome 18.  Further,
     a marker on chromosome 1 likely belongs on chromosome 18. 

<tr> <td align="right"><a href="plot.errorlod.html"><img
     src="thumb_plot.errorlod.jpg" alt="Plot of genotyping error LOD
  scores [23k]"></a>
     <td valign="top"><b>Genotyping error LOD scores</b>.  Colored pixels
     represent larger values of the error LOD scores described by
     Lincoln and Lander (1992) and which indicate likely genotyping
     errors.  The genotypes corresponding to purple pixels are likely
     to be in error. 

<tr> <td align="right"><a href="plot.geno.html"><img
     src="thumb_plot.geno.jpg" alt="Plot of genotype data with
  potential errors flagged [48k]"></a> 
     <td valign="top"><b>Observed marker genotype data</b> for an intercross.
     White, gray and black circles correspond to the genotypes AA, AB,
     and BB, respectively.  Genotypes flagged by red squares were
     indicated to be likely genotyping errors. 

<tr> <td align="right"> <a href="plot.info.html"><img
     src="thumb_plot.info.jpg" alt="Plot of proportion of missing
  genotype information [32k]"></a> 
     <td valign="top"><b>Proportion of missing genotype information</b>, calculated
     in two ways (using the concepts of entropy and of variance).
     Values of 0 correspond to complete data; large values
     indicate regions with little marker data.  The ticks below
     indicate the locations of genetic markers. 

<tr> <td align="right"> <a href="plot.scanone.html"><img
     src="thumb_plot.scanone.jpg" alt="Plot of LOD curves from
  genome scan [34k]"></a> <td valign="top"><b>LOD curves for a
     single-QTL model</b> for a selected set of chromosomes,
     calculated with the two-part model described by Boyartchuck et
     al. (2001), <i>Nat Genet</i> <b>27</b>, 259-260.  The ticks below
     indicate the locations of genetic markers. 

<tr> <td align="right"><a href="plot.scantwo.html"><img 
     src="thumb_plot.scantwo.jpg" alt="Plot of LOD scores from
  two-dimensional, two-QTL genome scan [132k]"></a> 
     <td valign="top"><b>LOD scores for a two-dimensional genome scan</b> with a
     two-QTL model, for a selected set of chromosomes.  Values in the
     lower right triangle are for joint two-locus LOD scores; values
     in the upper left triangle are LOD scores for the test of
     epistasis.  The color scale at the right indicates separate
     scales for the joint LOD scores (right) and the epistasis LOD
     scores (left). 

</table>

<p class=noindent><hr>

<p align="center">[ <a href="..">Home</a>
| <a href="../download">Download</a> | <a 
href="../faq">FAQ</a> | <a
href="../STATUS.txt">Status</a> | <a
href="../BUGS.txt">Bugs</a> |
<a href="../sampledata">Sample data</a> |
<a href="../tutorials">Tutorials</a> | 
<a href="http://www.rqtl.org/book">Book</a> |
<a href="../manual/qtl-manual.pdf">Manual</a> | <a
href="../citation.html">Citation</a> ] 

<script src="http://www.google-analytics.com/urchin.js" type="text/javascript">
</script>
<script type="text/javascript">
_uacct = "UA-609649-1";
urchinTracker();
</script> 
</body></html>
